### PR TITLE
bug 1255609 - Enable training on false positives

### PR DESCRIPTION
**Pull request by jwhitlock**
Add additional data to wiki's DocumentSpamAttempt:
* data - JSON-serialized data sent to Akismet
* review - Staff review status (default 'Needs Review')
* reviewer - Staff reviewer
* reviewed - Review date

Old DocumentSpamAttempts do not have this additional data, so their status is set to "Review Unavailable".

Update the Django admin for DocumentSpamAttempt so that it works better as a dashboard for second-guessing Akismet:
* Limit the length of title and slug on the list display
* Allow filtering by "Needs Review", to treat the admin as a review backlog.
* Add an editable dropdown to quickly review multiple edits from known users
* When a review of "Spam" is chosen, it confims Akismet's choice, and no futher API call is needed.
* When a review of "Ham / False Positive" is chosen, submit the original data to Akismet's submit_ham API endpoint. On failure, revert to unreviewed status so it can be tried again.

Add a management command and task to drop submission data for old records, saving database resources and removing unneeded PII.

Update the anonymization scripts to drop submission data.

It can be hard to test this locally. If you want to, it helps to create a user 'viagra-test-123', which will always be counted as spam:

spam dashboard

### CODE REVIEW

File: kuma/wiki/admin.py [hunk 1]
```diff
@@ -335,12 +341,81 @@ class DocumentZoneAdmin(admin.ModelAdmin):
...
+
+    def save_model(self, request, obj, form, change):
+        """If reviewed, set the reviewer, and submit ham as requested."""
+        send_ham = (obj.review == DocumentSpamAttempt.HAM and
+                    obj.reviewed is None and obj.data is not None)
+        if obj.review in (DocumentSpamAttempt.HAM, DocumentSpamAttempt.SPAM):
+            obj.reviewer = obj.reviewer or request.user
+            obj.reviewed = obj.reviewed or timezone.now()
+        if send_ham:
+            data = loads(obj.data)
```
Comment (jezdez) [inline kuma/wiki/admin.py]: mind using `json.loads` here to make it less ambiguous

File: kuma/wiki/admin.py [hunk 2]
```diff
@@ -335,12 +341,81 @@ class DocumentZoneAdmin(admin.ModelAdmin):
...
+            else:
+                message = 'Submitted ham for document spam attempt %s' % obj
+                self.message_user(request, message, level=messages.INFO)
+        obj.save()
+
+    SUBMISSION_NOT_AVAILABLE = "Akismet submission not available."
+
+    def submitted_data(self, instance):
+        if instance.data:
+            return dumps(loads(instance.data), indent=4)
```
Comment (jezdez) [inline kuma/wiki/admin.py]: `json...` here please

File: kuma/wiki/forms.py [hunk 3]
```diff
@@ -450,6 +451,7 @@ def akismet_error(self):
                 slug=self.cleaned_data['slug'],
                 user=self.request.user,
                 document=document,
+                data=dumps(self.akismet_parameters()),
```
Comment (jezdez) [inline kuma/wiki/forms.py]: `json.dumps` here

File: kuma/wiki/management/commands/delete_old_spam_attempt_data.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+class Command(BaseCommand):
+    help = "Delete old DocumentSpamAttempt.data"
+    option_list = BaseCommand.option_list + (
+        make_option('--days', dest="days", default=30, type=int,
+                    help="How many days 'old'"),
+    )
+
+    def handle(self, *args, **options):
+        self.options = options
```
Comment (jezdez) [inline kuma/wiki/management/commands/delete_old_spam_attempt_data.py]: No need to set `self.options` here

File: kuma/wiki/models.py [hunk 5]
```diff
@@ -1940,6 +1941,41 @@ class DocumentSpamAttempt(SpamAttempt):
...
+        blank=True,
+        null=True,
+    )
+
+    NEEDS_REVIEW = 0
+    HAM = 1
+    SPAM = 2
+    REVIEW_UNAVAILABLE = 3
+    REVIEW_CHOICES = (
+        (NEEDS_REVIEW, ugettext('Needs Review')),
```
Comment (jezdez) [inline kuma/wiki/models.py]: This should be using `ugettext_lazy` since it's evaluated at import time

File: kuma/wiki/models.py [hunk 6]
```diff
@@ -1940,6 +1941,41 @@ class DocumentSpamAttempt(SpamAttempt):
...
+        (HAM, ugettext('Ham / False Positive')),
+        (SPAM, ugettext('Confirmed as Spam')),
+        (REVIEW_UNAVAILABLE, ugettext('Review Unavailable')),
+    )
+    review = models.IntegerField(
+        choices=REVIEW_CHOICES,
+        default=NEEDS_REVIEW,
+        verbose_name=ugettext(
```
Comment (jezdez) [inline kuma/wiki/models.py]: needs to be lazy as well

File: kuma/wiki/models.py [hunk 7]
```diff
@@ -1940,6 +1941,41 @@ class DocumentSpamAttempt(SpamAttempt):
...
+            "Review of Akismet's classification as spam"),
+    )
+    reviewer = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        related_name='documentspam_reviewed',
+        blank=True,
+        null=True,
+        verbose_name=ugettext('Staff reviewer'),
```
Comment (jezdez) [inline kuma/wiki/models.py]: laaaazy

File: kuma/wiki/tests/test_admin.py [hunk 8]
```diff
@@ -2,21 +2,170 @@
...
+        "comment_author": "viagra-test-123",
+        "comment_author_email": "viagra-test-123@example.com"
+    }"""
+
+    class FakeMessageStorage(object):
+        def __init__(self):
+            self.storage = []
+
+        def add(self, level, message, extra_tags):
+            self.storage.append((level, message, extra_tags))
```
Comment (jezdez) [inline kuma/wiki/tests/test_admin.py]: there is a `get_messages` test method in `kuma.core.tests` to do something similar

File: kuma/wiki/tests/test_admin.py [hunk 9]
```diff
@@ -2,21 +2,170 @@
...
+        self.admin_user = User.objects.get(username='admin')
+        self.request = RequestFactory().get('/admin/dsa')
+        self.request.user = self.admin_user
+        self.request._messages = self.FakeMessageStorage()
+
+    def test_title_short(self):
+        dsa = DocumentSpamAttempt(title='A short title')
+        assert self.admin.title_short(dsa) == 'A short title'
+        dsa.title = (
+            'Call Usa 18884992506DELL printer 1 844 240 8386 support service'
```
Comment (jezdez) [inline kuma/wiki/tests/test_admin.py]: IDK, can you use a different title here, this seems not fair to help them with the google results

**Comment by jezdez** [comment]
A quick code review didn't show many issues, haven't tried it yet in practice tho

File: kuma/wiki/models.py (hunk 5 above)
Comment (jwhitlock) [inline kuma/wiki/models.py]: OK, changing these and the existing ``ugettext`` used above.

File: kuma/wiki/tests/test_admin.py (hunk 9 above)
Comment (jwhitlock) [inline kuma/wiki/tests/test_admin.py]: good point. changing to something boring.

File: kuma/wiki/tests/test_admin.py (hunk 8 above)
Comment (jwhitlock) [inline kuma/wiki/tests/test_admin.py]: and ``UserTestCase`` is derived from that, so I'll use that instead of yet another test class.

**Comment by jwhitlock** [comment]
Thanks for the quick review @jezdez. Force push to:

* Use ``json.loads`` and ``json.dumps`` instead of ``loads``, ``dumps``
* Use ``_`` (aka ``ugettext_lazy``) instead of ``ugettext`` in models
* Use faked message storage from ``kuma/core/tests/__init__.py``
* Don't feed the spammers
* Other cleanup from review

Akismet doesn't like a lot of our content, so we can test a lot of this in staging by making one-line changes.

File: kuma/wiki/admin.py [hunk 10]
```diff
@@ -335,12 +341,81 @@ class DocumentZoneAdmin(admin.ModelAdmin):
...
+        if send_ham:
+            data = json.loads(obj.data)
+            try:
+                self.submit_ham(request, data)
+            except (self.NotEnabled, AkismetError) as error:
+                obj.reviewer = None
+                obj.reviewed = None
+                obj.review = DocumentSpamAttempt.NEEDS_REVIEW
+                message = (
+                    'Unable to submit ham for document spam attempt %s: %s' %
```
Comment (robhudson) [inline kuma/wiki/admin.py]: Should this be localized? Or do we not localize in the admin?

File: kuma/wiki/management/commands/delete_old_spam_attempt_data.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+from django.core.management.base import BaseCommand
+from kuma.wiki.tasks import delete_old_documentspamattempt_data
+
+
+class Command(BaseCommand):
+    help = "Delete old DocumentSpamAttempt.data"
+    option_list = BaseCommand.option_list + (
+        make_option('--days', dest="days", default=30, type=int,
+                    help="How many days 'old'"),
```
Comment (robhudson) [inline kuma/wiki/management/commands/delete_old_spam_attempt_data.py]: Should this be in configman with most of the other settings?

File: kuma/wiki/models.py [hunk 12]
```diff
@@ -1937,9 +1938,43 @@ class DocumentSpamAttempt(SpamAttempt):
...
+    NEEDS_REVIEW = 0
+    HAM = 1
+    SPAM = 2
+    REVIEW_UNAVAILABLE = 3
+    REVIEW_CHOICES = (
+        (NEEDS_REVIEW, _('Needs Review')),
+        (HAM, _('Ham / False Positive')),
+        (SPAM, _('Confirmed as Spam')),
+        (REVIEW_UNAVAILABLE, _('Review Unavailable')),
+    )
```
Comment (robhudson) [inline kuma/wiki/models.py]: I'm unsuccessfully finding a pattern in how you use single quotes or double quotes. :grin:

**Comment by robhudson** [comment]
Finished a quick review. Asked some questions. Looks really nice.

File: kuma/wiki/management/commands/delete_old_spam_attempt_data.py (hunk 11 above)
Comment (jwhitlock) [inline kuma/wiki/management/commands/delete_old_spam_attempt_data.py]: I don't think management command's settings should be in configman. I copied the code from ``delete_old_revision_ps.py``, which is very similar. I'd need to look at others to see if any of them use configman.

Comment (robhudson) [inline kuma/wiki/management/commands/delete_old_spam_attempt_data.py]: Sorry, I meant django-constance. configman is something else entirely. :) You're closer to this than I am but I thought I'd ask.

File: kuma/wiki/admin.py (hunk 10 above)
Comment (jwhitlock) [inline kuma/wiki/admin.py]: Maybe someday. None of the existing code in kuma/wiki/admin.py is localized, so no, we don't today, other than the included localization in Django.

File: kuma/wiki/models.py (hunk 12 above)
Comment (jwhitlock) [inline kuma/wiki/models.py]: The pattern is double quotes, unless I notice I'm using them, and then I switch to single quotes. Single quotes aren't natural to me, and I'll be a lot more comfortable when [bug 1237977](https://bugzilla.mozilla.org/show_bug.cgi?id=1237977) is integrated.

Until then, I'll try to fix some of my double quotes and force push

**Comment by jwhitlock** [comment]
Force push to:
* Rebase on current master
* Change some double quotes to single quotes
* Fix ``./manage.py delete_old_spam_attempt_data``

**Comment by robhudson** [comment]
r+

**Pull request merged**